apps/emqx_bridge_kafka/src/emqx_bridge_kafka_impl_producer.erl

```diff
@@ -279,34 +303,54 @@ on_kafka_ack(_Partition, buffer_overflow_discarded, _Callback) ->
     %% do not apply the callback (which is basically to bump success or fail counter)
     ok.
 
-on_get_status(_InstId, #{client_id := ClientId, kafka_topic := KafkaTopic}) ->
+on_get_status(_InstId, #{client_id := ClientId} = State) ->
     case wolff_client_sup:find_client(ClientId) of
         {ok, Pid} ->
-            do_get_status(Pid, KafkaTopic);
+            do_get_status(Pid, State);
         {error, _Reason} ->
             disconnected
     end.
 
-do_get_status(Client, KafkaTopic) ->
+do_get_status(Client, #{kafka_topic := KafkaTopic} = State) ->
+    CheckIfTopicExists =
+        fun(Pid, Topic, Default) ->
+            case wolff_client:check_if_topic_exists(Pid, Topic) of
+                ok ->
+                    connected;
+                {error, unknown_topic_or_partition} ->
+                    {disconnected, State, unhealthy_target};
+                _ ->
+                    Default
+            end
+        end,
     %% TODO: add a wolff_producers:check_connectivity
     case wolff_client:get_leader_connections(Client, KafkaTopic) of
         {ok, Leaders} ->
-            %% Kafka is considered healthy as long as any of the partition leader is reachable
-            case
-                lists:any(
-                    fun({_Partition, Pid}) ->
-                        is_pid(Pid) andalso erlang:is_process_alive(Pid)
-                    end,
-                    Leaders
-                )
-            of
-                true ->
-                    connected;
-                false ->
-                    disconnected
-            end;
+            %% Kafka is considered healthy as long as any of the partition leader is reachable.
+            %% It's considered unhealthy if the target topic doesn't exist.
+            lists:foldl(
+                fun
+                    (_, connected) ->
+                        connected;
+                    ({_Partition, Pid}, Acc) ->
+                        case is_pid(Pid) andalso erlang:is_process_alive(Pid) of
+                            false ->
+                                Acc;
+                            true ->
+                                CheckIfTopicExists(Pid, KafkaTopic, Acc)
+                        end
+                end,
+                disconnected,
+                Leaders
+            );
```

Comment:
Maybe we could disentangle the two checks to make it more readable, and check the topic existence only once?

```erlang
on_get_status ->
    case do_get_status(..., State) of
         {disconnected, unhealthy} -> {disconnected, State, unhealthy};
         Status -> Status
    end.

%% ...

do_get_status ->
    HealthyLeaders = lists:filter(fun(L) -> is_pid ... end, Leaders),
    case HealthyLeaders of
        [Pid | _] ->
            do_get_topic_status(Pid, Topic);
        [] ->
            disconnected
    end.

do_get_topic_status ->
     try wolf_client:check_if_topic_exists(Pid, Topic) of
       ...
     catch
        _:_ -> disconnected
     end.
```